### PR TITLE
bases(buildd): update cloud.cfg to not reset apt's sources.list

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

When using a buildd base, `/etc/cloud/cloud.cfg` must be updated to prevent `/etc/apt/sources.list` from being reset when copying an instance.

This fixes an issue uncovered by https://github.com/snapcore/snapcraft/pull/4068

### CODE REVIEW

File: craft_providers/bases/buildd.py [hunk 1]
```diff
@@ -1015,6 +1016,20 @@ def _update_apt_sources(
                 details=errors.details_from_called_process_error(error),
             ) from error
 
+        # update cloud.cfg to prevent the sources.list file from being reset
+        try:
```
Comment (lengau) [inline craft_providers/bases/buildd.py:1020]: Is there any time we want this to succeed if `cloud.cfg` doesn't exist?

File: craft_providers/bases/buildd.py [hunk 2]
```diff
@@ -1015,6 +1016,20 @@ def _update_apt_sources(
...
+            _execute_run(
+                executor,
+                # 'aapt' is not a typo, the first 'a' is the sed command to append
+                # this is a shlex-compatible way to append to a file
+                ["sed", "-i", "$ aapt_preserve_sources_list: true", cloud_config],
```
Comment (lengau) [inline craft_providers/bases/buildd.py:1025]: if cloud.cfg exists but is empty, this will exit with status 0 but not actually append to the file.

**Review by lengau — APPROVED** [review]
I've noted two potential edge cases. If we don't need to worry about them then I'm happy with this as is.

File: craft_providers/bases/buildd.py (hunk 1 above)
Comment (mr-cal) [reply craft_providers/bases/buildd.py:1033]: Great idea.  `cloud.cfg` (from `cloud-init)`is fundamental to buildd images, so I decided to raise an error if it doesn't exist.

File: craft_providers/bases/buildd.py (hunk 2 above)
Comment (mr-cal) [reply craft_providers/bases/buildd.py:1038]: Thanks, I added a check to make sure `cloud.cfg` exists and is not empty.

**Review by lengau — APPROVED** [review]

**Pull request merged**